### PR TITLE
PARQUET-1819: [C++] Refactor decoding

### PR DESCRIPTION
**Pull request by pitrou**
Also add an additional size check before reading the length of a byte array.

### CODE REVIEW

**Comment by pitrou** [comment]
cc @fsaintjacques @wesm

**Comment by fsaintjacques** [comment]
Could you keep this in the same file to simplify the review, and once it's approved, extract the decoding part in a new file in a subsequent commit? I'm not in the mood to read 1500 lines of parquet decoding.

**Comment by pitrou** [comment]
Ok, the diff should be much shorter now.

**Comment by pitrou** [comment]
AppVeyor build: https://ci.appveyor.com/project/pitrou/arrow/builds/31651804

**Comment by pitrou** [comment]
@fsaintjacques Would be nice if you could give this a quick review, so that I can move forward with #6690.

File: cpp/src/parquet/encoding.cc
Comment (bkietz) [inline cpp/src/parquet/encoding.cc]:
```suggestion
  const int64_t consumed_length = static_cast<int64_t>(4 + len);
```

File: cpp/src/arrow/visitor_inline.h
Comment (bkietz) [inline cpp/src/arrow/visitor_inline.h]: This doesn't seem worthwhile given that we have `internal::VisitBits` and `internal::VisitBitsUnrolled`. The Status return signature is not useful in `parquet::` since we can throw an exception to effect early termination

File: cpp/src/parquet/encoding.cc
Comment (bkietz) [inline cpp/src/parquet/encoding.cc]: Coding style: please reintroduce the constant (named `kPrefixSize` maybe) or use 
```suggestion
        if (ARROW_PREDICT_FALSE(len_ < sizeof(int32_t))) {
```
rather than `4`

File: cpp/src/parquet/encoding.cc
Comment (bkietz) [inline cpp/src/parquet/encoding.cc]: :+1:

File: cpp/src/parquet/encoding.cc
Comment (bkietz) [inline cpp/src/parquet/encoding.cc]:
```suggestion
  len_ -= kNumStreams * values_decoded;
```

**Review by bkietz — CHANGES_REQUESTED** [review]
This looks good, just a few comments

File: cpp/src/parquet/encoding.cc
Comment (pitrou) [reply cpp/src/parquet/encoding.cc]: -1. `sizeof(int32_t)` is pointless pedantry.

File: cpp/src/arrow/visitor_inline.h
Comment (pitrou) [reply cpp/src/arrow/visitor_inline.h]: Well, this function also gracefully separates the case where `null_count` is 0 (and the null bitmap pointer potentially null), both for correctness and better performance.

File: cpp/src/parquet/encoding.cc
Comment (pitrou) [reply cpp/src/parquet/encoding.cc]: Fair enough.

File: cpp/src/parquet/encoding.cc
Comment (pitrou) [reply cpp/src/parquet/encoding.cc]: `sizeof(T)` is logically more exact, though it's the same value (the length in bytes is decreased by the number of values decoded times the value size).

File: cpp/src/parquet/encoding.cc
Comment (fsaintjacques) [inline cpp/src/parquet/encoding.cc]: Not related, to your change, but `dict_values[index]` should check for the index bounds.

File: cpp/src/parquet/encoding.cc
Comment (fsaintjacques) [inline cpp/src/parquet/encoding.cc]: Ditto with bounds.

File: cpp/src/parquet/encoding.cc
Comment (fsaintjacques) [inline cpp/src/parquet/encoding.cc]: Ditto.

**Review by fsaintjacques — CHANGES_REQUESTED** [review]

**Review by wesm — COMMENTED** [review]
Thanks for doing this long overdue improvement. I don't have anything major to add, but I'm going to check the benchmarks locally

**Comment by wesm** [comment]
Benchmarks show no perceptible difference based on a quick glance

File: cpp/src/parquet/encoding.cc
Comment (pitrou) [reply cpp/src/parquet/encoding.cc]: That sounds dubious to me.

File: cpp/src/parquet/encoding.cc
Comment (pitrou) [reply cpp/src/parquet/encoding.cc]: Will do, thank you.

**Comment by pitrou** [comment]
+1, will merge if CI green.

**Pull request closed without merging**
